Fix/workaround for wxLocale breaking Unicode string formatting

lanurmi opened the pull request:
Also attempt to test whether wxLocale breaks formatting Unicode strings.
See #23450.

lanurmi committed:
Attempt to test whether wxLocale breaks formatting Unicode strings

lanurmi committed:
Fix/workaround for wxLocale breaking Unicode string formatting

vadz commented on src/common/intl.cpp:
```diff
@@ -396,6 +396,12 @@ bool wxLocale::DoCommonPostInit(bool success,
     }
 
+#if defined(__WXOSX__)
+    // LC_CTYPE is set in DoCommonPreInit() (see init.cpp). Set it again here,
```
I think we might want have some `wxLocale::OSXFixCRT()` or something like this and call it from both places, even if it's a trivial one-line function because this would allow documenting this issue in a single place -- and hopefully get rid of it easier in the future, too.

vadz replied:
I've done this now, please let me know if you have any objections.

vadz reviewed (commented):
Thanks for the fix, we should, of course, do this if it doesn't work without it even if it's somewhat unexpected for `LC_CTYPE` to "change on its own" like this (I'd like to document it, but I'm not sure where).

I think this should also be backported to 3.2?

lanurmi commented:
> Thanks for the fix, we should, of course, do this if it doesn't work without it even if it's somewhat unexpected for `LC_CTYPE` to "change on its own" like this (I'd like to document it, but I'm not sure where).

It is somewhat unexpected, though it has been changing like that already in 2.9.something. From the caller's perspective, this fix prevents it from changing again and retains old behaviour.

> I think this should also be backported to 3.2?

Absolutely, this is an issue that causes subtle bugs with random strings missing for no apparent reason or pattern.

vadz committed:
Add wxEnsureLocaleIsCompatibleWithCRT() helper

No real changes, just refactor the code to use this function instead of
calling setlocale(LC_CTYPE) twice.

This makes it more clear what we're doing and will make it simpler to
get rid of this when we don't have to do it any longer.

vadz committed:
Generalize wxOSX locale workaround to all ports under Darwin

As all ports use the same CRT, it makes more sense to check for Darwin,
rather than wxOSX, in wxEnsureLocaleIsCompatibleWithCRT(), as other
ports such as wxGTK would also be affected by this problem if they were
to be used under macOS.
